Extensions/GDevelopAuthentication/gdevelopauthenticationtools.ts

```diff
@@ -0,0 +1,865 @@
+      _authenticationBanner = computeNotAuthenticatedBanner(runtimeScene);
+    };
+
+    /**
+     * Retrieves the user information from the local storage, and store
+     * them in the extension variables.
+     * Returns if the fetched user is different than the current one used
+     * in the variables. (useful when the user changes)
+     */
+    const fetchAuthenticatedUserFromLocalStorage = (): boolean => {
+      const _gameId = gdjs.projectData.properties.projectUuid;
+
+      const authenticatedUserStorageItem = window.localStorage.getItem(
+        `${_gameId}_authenticatedUser`
+      );
+      if (!authenticatedUserStorageItem) {
+        _checkedLocalStorage = true;
+        return _userId === null;
+      }
+      const authenticatedUser = JSON.parse(authenticatedUserStorageItem);
+      const hasUserChanged = authenticatedUser.userId !== _userId;
+
+      _username = authenticatedUser.username;
+      _userId = authenticatedUser.userId;
+      _userToken = authenticatedUser.userToken;
+      _checkedLocalStorage = true;
+
+      return hasUserChanged;
+    };
+
+    /**
+     * Reads the event sent by the authentication window and
+     * display the appropriate banner.
+     */
+    const receiveMessageFromAuthenticationWindow = function (
+      runtimeScene: gdjs.RuntimeScene,
+      event: MessageEvent,
+      { checkOrigin }: { checkOrigin: boolean }
+    ) {
+      const _gameId = gdjs.projectData.properties.projectUuid;
+      const allowedOrigin = 'https://liluo.io';
+
+      // Check origin of message.
+      if (checkOrigin && event.origin !== allowedOrigin) {
+        throw new Error(`Unexpected origin: ${event.origin}`);
+      }
+      // Check that message is not malformed.
+      if (!event.data.id) {
+        throw new Error('Malformed message');
+      }
+
+      // Handle message.
+      switch (event.data.id) {
+        case 'authenticationLoaded': {
+          if (_authenticationIframe) {
+            // Hide loader.
+            if (_authenticationSubContainer && _authenticationLoader) {
+              _authenticationSubContainer.removeChild(_authenticationLoader);
+            }
+            // Show iframe
+            _authenticationIframe.style.opacity = '1';
+          }
+          break;
+        }
+        case 'authenticationResult': {
+          if (!(event.data.body && event.data.body.token)) {
+            throw new Error('Malformed message.');
+          }
+          const username = event.data.body.username;
+          if (!username) {
+            logger.warn('The authenticated player does not have a username');
+          }
+          window.localStorage.setItem(
+            `${_gameId}_authenticatedUser`,
+            JSON.stringify({
+              username: _username,
+              userId: _userId,
+              userToken: _userToken,
+            })
+          );
+          _username = username;
+          _userId = event.data.body.userId;
+          _userToken = event.data.body.token;
```

Comment:
I feel like those 3 lines should be put above the local storage item setting instruction.

Reply:
👍